tests/shared/src/test/scala/cats/effect/std/SupervisorSpec.scala

```diff
@@ -229,5 +235,68 @@ class SupervisorSpec extends BaseSpec {
       // if this doesn't work properly, the test will hang
       test.start.flatMap(_.join).as(ok).timeoutTo(2.seconds, IO(false must beTrue))
     }
+
+    "supervise / finalize race" in real {
+      superviseFinalizeRace(constructor(false, None), IO.never[Unit])
+    }
+
+    "supervise / finalize race with checkRestart" in real {
+      superviseFinalizeRace(constructor(false, Some(_ => true)), IO.canceled)
+    }
+
+    def superviseFinalizeRace(mkSupervisor: Resource[IO, Supervisor[IO]], task: IO[Unit]) = {
+      val tsk = IO.uncancelable { poll =>
+        mkSupervisor.allocated.flatMap {
+          case (supervisor, close) =>
+            supervisor.supervise(IO.never[Unit]).replicateA(100).flatMap { fibers =>
+              val tryFork = supervisor.supervise(task).map(Some(_)).recover {
+                case ex: IllegalStateException
+                    if ex.getMessage == "supervisor already shutdown" =>
+                  None
```

Comment:
Might be a bit easier to diagnose an eventual failure here if we convert this into a side-effecting test assertion.

```suggestion
                case ex: IllegalStateException =>
                  ex.getMessage mustEqual "supervisor already shutdown"
                  None
```